### DISCUSSION TITLE
Why Wagtail consume average 5ms per request for query site info, how to avoid it ?

### DISCUSSION THREAD

**Discussion by dusty-cjh** [wagtail/wagtail#9701] [General]
hi, I'm a beginner of wagtail, I'm curious about something when I try the bakery demo.

here is the debug_toolbar output from [bakery demo](https://github.com/wagtail/bakerydemo):
![image](https://user-images.githubusercontent.com/45847340/203040701-bdc032f3-f85c-4f90-9995-62582eddb306.png)

I'm concerned about why wagtail queries site info with super complicated SQL in every page request, is that possible to add a cache or assign some fixed data?

**Comment by RealOrangeOne** [comment]
This is part of the query Wagtail uses to determine which site is being used for a given request, and by extension what the root page is to start resolving from.

I did some work a little while ago to try and cache and simplify this query to improve performance: https://github.com/wagtail/wagtail/pull/8937/, after which yes it'll definitely be possible to cache it.

**Comment by allcaps** [comment]
Why?

Optimization can prove to have been a waste of time if the optimized code is only a small part of the workload. Making a routine 10 times faster doesn't help much if that routine is only consuming 1% of the running time to begin with and it may result in more complicated and buggier code.

Caching comes at a cost: more complicated code, and the overhead of caching framework itself.

What about caching all queries and/or cache your complete application. Eg https://django-cachalot.readthedocs.io/en/latest/ OR https://varnish-cache.org/ OR a CDN.

**Reply by RealOrangeOne** [reply]
This is true, but reducing page loads by 5ms is still a benefit. Especially something which is so highly cacheable. Depending on what cache is used will depend on the performance gains, but I think they're still worth it. The code isn't the simplest, so it'd be nice to cache it and avoid running it on _every_ request.

**Reply by SebCorbin** [reply]
I agree with the comment but I hope we come to a situation like in Django core where `django.contrib.sites` is not required for Django to run because for most use cases of Wagtail, there is just one site.

**Reply by RealOrangeOne** [reply]
Extracting the multi-site functionality from Wagtail could indeed be a nice performance bump. But that's almost certainly going to be a very large piece of work - lots of Wagtail internally assumes there might be multiple sites, and entire features are built around it (eg site settings).
